rfcs/049-catalogue-api-aggregations-modelling.md

```diff
@@ -0,0 +1,176 @@
+
+The API uses [Elasticsearch aggregations][es_aggs] to aggregate over a field in internal model.
+Elasticsearch will return single string values, which the API then serialises into the display model.
+
+For example, our language aggregation starts as [an ES terms aggregation][terms_agg] over the `data.languages.id` field:
+
+```http
+GET /works-indexed-2022-04-28/_search
+{
+  "aggs": {
+    "languages": {
+      "terms": { "field": "data.languages.id" }
+    }
+  }
+}
+```
+
+The Elasticsearch API response can only use a single string in its buckets, for example:
+
+```
+{
+  "aggregations" : {
+    "languages" : {
+      "buckets" : [
+        {
+          "key" : "eng",
+          "doc_count" : 691840
+        },
+        {
+          "key" : "fre",
+          "doc_count" : 67187
+        },
+        ...
+```
+
+The API has to contain enough internal/display model logic to interpret these values -- to know that, say, `eng` means English and it's serialised as id/label/type.
+This is precisely the sort of model coupling we're trying to get away from.
+
+For more complex types, we've had to jump through hoops to shoehorn aggregations into this approach -- e.g. contributor values are stored like `person:Henry Wellcome` because we need both the type and the label to return them correctly.
+
+It would be nice if we could remove this coupling and simplify how aggregations work in the API.
+
+[es_aggs]: https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations.html
+[terms_agg]: https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-terms-aggregation.html
+
+
+
+## Proposed solution
+
+We add a new field `query.aggregatableValues` to the Elasticsearch model.
+This would be an object field in which the values are lists of strings:
+
+```
+type Query {
+  aggregatableValues: Map[String, List[String]]
+  ...
+}
+```
+
+The labels are the aggregation types (languages, work types, licenses) and the values are blobs of display JSON stored as strings.
+That is, we create the display JSON, then serialise the whole object as a string, and store that.
+
+This is easiest to understand with an example:
+
+```json
+{
+  "id": "example-work",
+  "query": {
+    "aggregatableValues": {
+      "languages": [
+        " { \"id\" : \"eng\", \"label\": \"English\", \"type\": \"Language\" } ",
+        " { \"id\" : \"fre\", \"label\": \"French\", \"type\": \"Language\" } "
+      ],
+      "items.locations.license": [
+        " { \"id\": \"pdm\", \"label\": \"Public Domain Mark\", \"url\": \"https://creativecommons.org/share-your-work/public-domain/pdm/\", \"type\": \"License\" } "
+      ],
+      ...
+    }
+  },
+  ...
+}
+```
+
+The ingestors would populate these `aggregatableValues` fields when it indexed a work.
+This would be mapped as a `keyword` field in Elasticsearch.
+
+The API would aggregate over these fields specifically, and copy the values into the `data` field of our aggregation buckets.
```

Comment:
So in the API, we end up with code that transforms:
```{
           "key" : "blob of JSON as a string",
           "doc_count" : 691840
         },```
into 
```
{
         "data": that same JSON as an object
         "count": 691840,
         "type": "AggregationBucket"
       },
```
essentially ending up with something like this in the API code (excuse the Pythoning)
```
return [ 
  {
     "data: "json.load(bucket["key"]),
     "count": bucket["doc_count"],
     "type": "AggregationBucket"
   }
  for bucket in es_buckets
]
```
which is pretty simple.

Reply:
Yes, exactly. (I had something pretty close to that while testing locally.)

Would it be helpful to add something like your comment to the RFC?